File: examples/runnable/06-testing/python/reminder.py
"""Reminder workflow under test: sleep a day, then return a message.

The Python equivalent of the Java ReminderWorkflow testing lab. The point is that
a one-day sleep completes in milliseconds under the time-skipping test
environment, so the test runs fast and hermetically — no server, no real wait.
"""

from datetime import timedelta

from temporalio import activity, workflow

TASK_QUEUE = "test-reminder"


@activity.defn
async def lookup_email(user_id: str) -> str:
    # Real impl would hit a user service; tests register a fake in its place.
    raise NotImplementedError


@workflow.defn
class ReminderWorkflow:
    @workflow.run
    async def remind_after_one_day(self, message: str) -> str:
        await workflow.sleep(timedelta(days=1))  # skipped by the test env
        return f"Reminder: {message}"


@workflow.defn
class EmailReminderWorkflow:
    @workflow.run
    async def remind(self, user_id: str) -> str:
        email = await workflow.execute_activity(
            lookup_email, user_id, start_to_close_timeout=timedelta(seconds=10)
        )
        return f"sent to {email}"

File: examples/runnable/06-testing/go/reminder.go
// Package main holds the reminder testing lab: a Workflow that sleeps a day then
// returns a message, plus an Activity-backed variant. The Go equivalent of the
// Java ReminderWorkflow testing lab. The testsuite env (see reminder_test.go)
// runs these in-process and skips time, so the one-day sleep costs nothing.
package main

import (
	"context"
	"time"

	"go.temporal.io/sdk/workflow"
)

const TaskQueue = "test-reminder"

// LookupEmail is an Activity. The real impl would hit a user service; the test
// mocks it.
func LookupEmail(ctx context.Context, userID string) (string, error) {
	return "", nil
}

// ReminderWorkflow sleeps a day (skipped by the test env), then returns.
func ReminderWorkflow(ctx workflow.Context, message string) (string, error) {
	if err := workflow.Sleep(ctx, 24*time.Hour); err != nil {
		return "", err
	}
	return "Reminder: " + message, nil
}

// EmailReminderWorkflow calls an Activity, so a test can mock it.
func EmailReminderWorkflow(ctx workflow.Context, userID string) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
	})
	var email string
	if err := workflow.ExecuteActivity(ctx, LookupEmail, userID).Get(ctx, &email); err != nil {
		return "", err
	}
	return "sent to " + email, nil
}
